### PR TITLE
Consensus folk detection

### PR DESCRIPTION
**Pull request by halfprice**
## Description 

This PR implements consensus folk detection by adding consensus output digest into consensus commit prologue transaction. The digest implementation uses the existing ConsensusOutputCommit.

## Test Plan 

Unit test to test to test basic ConsensusCommitPrologueV2
Unit test to test CheckpointSummary creates different digest if the included ConsensusCommitPrologueV2 transactions have different consensus output digest.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Comment by halfprice** [comment]
Send out the current version while I'm still evaluating the testing plan.

The PR looks big, but the core logic is quite simple. Please see below notes.

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs [hunk 1]
```diff
@@ -1576,6 +1576,14 @@ impl From<&TransactionKind> for TransactionBlockKind {
                 };
                 TransactionBlockKind::ConsensusCommitPrologueTransaction(consensus)
             }
+            TransactionKind::ConsensusCommitPrologueV2(x) => {
+                let consensus = ConsensusCommitPrologueTransaction {
+                    epoch_id: x.epoch,
+                    round: Some(x.round),
+                    timestamp: DateTime::from_ms(x.commit_timestamp_ms as i64),
+                };
+                TransactionBlockKind::ConsensusCommitPrologueTransaction(consensus)
```
Comment (halfprice) [inline crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:1585]: I currently didn't create a ConsensusCommitPrologueV2Transaction in graphql rpc. Not sure if we need to (since the digest is mainly for internal purpose? But I do see a need to show it in query for debugging purpose). Even if we do, I think I need to do it in a separate PR given that the amount of work to make graph ql work with digest is non trivial (at least based on my current attempt).

File: crates/sui-core/src/consensus_handler.rs [hunk 2]
```diff
@@ -267,7 +268,16 @@ impl<T: ObjectStore + Send + Sync, C: CheckpointServiceNotify + Send + Sync>
...
+        let consensus_digest = match self
+            .epoch_store
+            .protocol_config()
+            .include_consensus_digest_in_prologue()
+        {
+            true => Some(consensus_output.consensus_digest()),
+            false => None,
+        };
+        let prologue_transaction =
+            self.consensus_commit_prologue_transaction(round, timestamp, consensus_digest);
```
Comment (halfprice) [inline crates/sui-core/src/consensus_handler.rs:280]: This change block is basically the core change of this PR

File: crates/sui-types/src/messages_checkpoint.rs [hunk 3]
```diff
@@ -771,4 +773,67 @@ mod tests {
...
+            None,
+            GasCostSummary::default(),
+            None,
+            100,
+        )
+    }
+
+    // Tests that ConsensusCommitPrologue with different consensus commit digest will result in different checkpoint content.
+    #[test]
+    fn test_checkpoing_summary_with_different_consensus_digest() {
```
Comment (halfprice) [inline crates/sui-types/src/messages_checkpoint.rs:800]: This test is the best scenario that I can think of right now to test the current code without writing a complex integration test with artificial folk. It basically tests that given a different consensus output digest, the checkpoint digest will be different. Any other testing suggestions are appreciated!

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs (hunk 1 above)
Comment (amnn) [reply crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:1585]: It's fine to leave it as-is, thanks @halfprice, but could you add a TODO comment? I'll pick it up and add the digest after you've landed this change.

File: crates/sui-types/src/transaction.rs [hunk 4]
```diff
@@ -2349,14 +2371,25 @@ impl VerifiedTransaction {
         epoch: u64,
         round: u64,
         commit_timestamp_ms: CheckpointTimestamp,
+        consensus_digest: Option<ConsensusCommitDigest>,
```
Comment (lxfind) [inline crates/sui-types/src/transaction.rs:2374]: No strong opinion, but I think it might look clearer if we have two separate constructors, for v1 and v2 respectively, instead of relying on the None-ness of the argument.

Comment (halfprice) [reply crates/sui-types/src/transaction.rs]: Done. I didn't separate the two constructors because we need to make multiple `v2` functions from the call site to here. But I can also see the benefit to separate V1 and V2 from the call site of generating the consensus prologue transaction.

**Review by mystenmark — APPROVED** [review]

**Comment by mystenmark** [comment]
Looks great to me!

**Review by mwtian — APPROVED** [review]
Looks good!

File: crates/sui-core/src/unit_tests/authority_tests.rs [hunk 5]
```diff
@@ -3039,6 +3040,43 @@ async fn test_refusal_to_sign_consensus_commit_prologue() {
     ));
 }
 
+#[tokio::test]
+async fn test_refusal_to_sign_consensus_commit_prologue_v2() {
```
[the remark this answers is not in the document]
Comment (halfprice) [reply crates/sui-core/src/unit_tests/authority_tests.rs:3044]: Correct

File: narwhal/types/src/consensus.rs [hunk 6]
```diff
@@ -417,10 +417,18 @@ impl CommittedSubDagShell {
 /// Shutdown token dropped when a task is properly shut down.
 pub type ShutdownToken = mpsc::Sender<()>;
 
-// Digest of ConsususOutput and CommittedSubDag
-#[derive(Clone, Copy, Default, PartialEq, Eq, Hash, PartialOrd, Ord)]
+// Digest of ConsususOutput and CommittedSubDag.
+// In non-byzantine environment, ConsensusOutputDigest of each consensus output in different
```
[the remark this answers is not in the document]
Comment (halfprice) [reply narwhal/types/src/consensus.rs:421]: Done

File: crates/sui-types/src/digests.rs [hunk 7]
```diff
@@ -922,3 +922,58 @@ impl ZKLoginInputsDigest {
...
+
+impl From<[u8; 32]> for ConsensusCommitDigest {
+    fn from(digest: [u8; 32]) -> Self {
+        Self::new(digest)
+    }
+}
+
+impl fmt::Display for ConsensusCommitDigest {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&self.0, f)
```
[the remark this answers is not in the document]
Comment (halfprice) [reply crates/sui-types/src/digests.rs:969]: This requires adding base64 as the package dependency. Is that ok?

[the remark this answers is not in the document]
Comment (mwtian) [reply crates/sui-types/src/digests.rs:969]: I believe it should be ok, since base64 is pretty low level.

[the remark this answers is not in the document]
Comment (halfprice) [reply crates/sui-types/src/digests.rs]: Thanks for confirming! Done

File: crates/sui-graphql-e2e-tests/tests/transactions/system.exp [hunk 8]
```diff
@@ -253,7 +253,7 @@ Response: {
     "transactionBlockConnection": {
       "nodes": [
         {
-          "digest": "J2gfeoDEySDWfUEdboMehk45L9ndHHFv3nLF16UWXxej",
+          "digest": "CHyKA8xirNCGtcDMLJcxBZsxGyAA7DNrzDEVodxZzNn",
```
Comment (halfprice) [inline crates/sui-graphql-e2e-tests/tests/transactions/system.exp:256]: Hi @amnn , can you please take a look if updating these digest is correct? The old one didn't include the ConsensusOutputDigest, and since we are including it, the digest here needs to change.
